text/0000-build-std/5-stage-1b.md

```diff
@@ -0,0 +1,986 @@
+added manually by users, however:
+
+- The standard library's dependencies are not part of the stable interface of
+  the standard library and it is not desirable that users can observe their
+  existence or depend on them directly
+
+- Other crates in the sysroot added by users are not something that can
+  reasonably be supported by build-std and these crates should become regular
+  dependencies
+
+↩ [*Proposal*][proposal]
+
+## Why allow all names for `builtin` crates on nightly?
+[rationale-nightly-builtin-crates]: #why-allow-all-names-for-builtin-crates-on-nightly
+
+For any crate shipped with the standard library in the sysroot, the user can
+already write an `extern crate` declaration to use it. All crates other than
+`std`, `alloc` or `core` are marked unstable either explicitly or implicitly
+with the use of `-Zforce-unstable-if-unmarked` so this does not allow items from
+these crates to be used on stable.
+
+For example, some users write benchmarks using `libtest` and have written
+`extern crate test` without the `#[cfg(test)]` attribute to load the crate.
+There may be other niche uses of unstable sysroot crates that this enables to
+continue on nightly toolchains.
+
+All names are permitted for `builtin` crates rather than an allowlist to avoid
+Cargo needing to hardcode the names of many of the crates in the sysroot, which
+are inherently unstable.
+
+↩ [*Proposal*][proposal]
+
+## Why not migrate to always requiring explicit standard library dependencies?
+[rationale-no-migration]: #why-not-migrate-to-always-requiring-explicit-standard-library-dependencies
+
+Explicit standard library dependencies with `builtin = true` will necessarily
+only be understood by newer versions of Cargo.
+
+If all packages were required to add explicit dependencies (perhaps over an
+edition or through some other mechanism), then every crate would require the
+newest version of Cargo to be understood, effectively raising the MSRV of every
+Rust crate.
+
+If only `no_std` crates (or crates with a `std` feature) add explicit
+dependencies on `core` or `alloc` then a much smaller percentage of the crates
+ecosystem will require the newest Cargo versions for their new explicit standard
+library dependencies to be understood.
+
+Alternative syntaxes, such as requiring `version = "*"` for explicit standard
+library dependencies, could be worthwhile to maintain a greater level of
+compatibility with older toolchain versions. Any currently accepted syntax would
+necessarily be interpreted differently by the build-std-supporting versions of
+Cargo, so this approach has its own complications. For example, while
+`version = "*"` would be understood by older versions of Cargo, it would attempt
+to find the standard library crates on crates.io and fail unless empty crates
+were published named `core`, `alloc` and `std`. This is not a build-std specific
+issue and is true of any RFC adding to what can be written in `Cargo.toml`.
```

Comment:
I'm a bit surprised `version = "*"` even comes up.  That is syntax for registry sources and would be confusing to users and make life more difficult in cargo.  It also offers negligable benefits here. It would be understood by old cargo's but still not usable.

If we were to talk about Index entry impact, what this would mean is that older cargos would resolve dependencies as if 
- Option 1: the package version with `core.builtin = true` doesn't exist but the error would say it exists and list the `package.rust-version` (if set).  This is if we make the schema incompatible
- Option 2: the package version would resolve fine and the build would fail.  This is if we extend the schema with new fields (a `deps2` field like our `features2`)

Reply:
I'm not sure I understand your options around index entries, but I've changed this paragraph to something more general and watertight in https://github.com/davidtwco/rfcs/pull/2/commits/9355da7e244ef8623c088d5fb9d31e5c5f2caa7c

Reply:
The focus on

> were considered to maintain a greater level of compatibility with older toolchain versions.

feels off.  There are two tiers of being compatible with older toolchains
- Index entry
- The manifest


We cannot maintain compatibility with older toolchains because the manifest will inherently be incompatible.

We can maintain compatibility at the Index entry level which is fairly narrow in meaning because its not like the entry can be used in any meaningful way.  This is done for the quality of error reporting with older Cargos that would treat certain index entry issues as the index entry is non-existent (which has been fixed for a while).

Reply:
I've incorporated this into another rework of this in https://github.com/davidtwco/rfcs/pull/2/commits/39c0f05e9ef159985b20529a3a2cf5b4f6ea44f2